site/source/docs/optimizing/Module-Splitting.rst

```diff
@@ -0,0 +1,368 @@
+produces an application.wasm.orig file. application.wasm.orig is the original,
+unmodified module that a normal Emscripten build would produce, while
+application.wasm has been instrumented by wasm-split to collect profiles.
+
+The instrumented module has an additional exported function,
+``__write_profile``, that takes as arguments a pointer and length for an
+in-memory buffer to which it will write the profile. ``__write_profile`` returns
+the length of the profile, and only writes the data if the supplied buffer is
+large enough. ``__write_profile`` can be called externally from JS or
+internally, from the application itself. For simplicity, we will just call it at
+the end of our main function here, but note that this will mean that any
+functions called after main, such as destructors for global objects, will not be
+included in the profile.
+
+Here’s the function to write the profile and our new main function::
+
+  EM_JS(void, write_profile, (), {
+    var __write_profile = Module['asm']['__write_profile'];
+    if (__write_profile) {
+
+      // Get the size of the profile and allocate a buffer for it.
+      var len = __write_profile(0, 0);
+      var ptr = _malloc(len);
+
+      // Write the profile data to the buffer.
+      __write_profile(ptr, len);
+
+      // Write the profile file.
+      var profile_data = new Uint8Array(buffer, ptr, len);
+      nodeFS.writeFileSync('profile.data', profile_data);
+
+      // Free the buffer.
+      _free(ptr);
+    }
+  });
+
+  int main() {
+    int i = get_number();
+    if (i == 0) {
+      foo();
+    } else if (i == 1) {
+      bar();
+    } else {
+      unsupported(i);
+    }
+    write_profile();
+  }
+
+Note that we only try to write the profile if the ``__write_profile`` export
+exists. This is important because only the instrumented, unsplit module exports
+``__write_profile``. The split modules will not include the profiling
+instrumentation or this export.
+
+Our new write_profile function depends on malloc and free being available to JS,
+so we need to explicitly export them on the command line::
+
+  $ emcc application.c -o application.js -sSPLIT_MODULE -sEXPORTED_FUNCTIONS=_malloc,_free,_main
+
+Now we can run our application, which produces a profile.data file. The next
+step is to use wasm-split and the profile to split the original module,
+application.wasm::
+
+  $ wasm-split --enable-mutable-globals --export-prefix=% application.wasm.orig -o1 application.wasm -o2 application.deferred.wasm --profile=profile.data
+
+Let’s break down what all those options are for.
+
+``--enable-mutable-globals``
+  This option enables the mutable-global target feature, which allows mutable
+  Wasm globals (as opposed to C/C++ globals) to be imported and exported.
+  wasm-split has to share mutable globals between the primary and secondary
+  modules, so it requires this feature to be enabled.
+
+``--export-prefix=%``
```

Comment:
is this required? can we just make `%` the default?

Reply:
Yeah, we could update wasm-split to make that the default. Right now the default is empty string.